pm/recipient.go

```diff
@@ -0,0 +1,207 @@
+package pm
+
+import (
+	"crypto/hmac"
+	"crypto/rand"
+	"crypto/sha256"
+	"math/big"
+	"sync"
+
+	ethcommon "github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/pkg/errors"
+)
+
+var (
+	errInvalidRecipientRandFromSeed = errors.New("invalid recipientRand generated from seed")
+	errInvalidRecipientRandRevealed = errors.New("invalid already revealed recipientRand")
+
+	errInvalidTicketFaceValue   = errors.New("invalid ticket faceValue")
+	errInvalidTicketWinProb     = errors.New("invalid ticket winProb")
+	errInvalidTicketSenderNonce = errors.New("invalid ticket senderNonce")
+
+	errZeroDepositAndPenaltyEscrow = errors.New("sender has zero deposit and penalty escrow")
+)
+
+// Recipient is an interface which describes an object capable
+// of receiving tickets
+type Recipient interface {
+	// ReceiveTicket validates and processes a received ticket
+	ReceiveTicket(ticket *Ticket, sig []byte, seed *big.Int) (won bool, err error)
+
+	// RedeemWinningTicket redeems a winning ticket with the broker
+	RedeemWinningTicket(ticketID ethcommon.Hash) error
+
+	// TicketParams returns the recipient's currently accepted ticket parameters
+	// for a provided sender ETH adddress
+	TicketParams(sender ethcommon.Address) (*TicketParams, error)
+}
+
+// recipient is an implementation of the Recipient interface that
+// receives tickets and redeems winning tickets
+type recipient struct {
+	val    Validator
+	broker Broker
+	store  TicketStore
+
+	secret [32]byte
+
+	invalidRands sync.Map
+
+	senderNonces     map[string]uint64
+	senderNoncesLock sync.Mutex
+
+	faceValue *big.Int
+	winProb   *big.Int
+}
+
+// NewRecipient creates an instance of a recipient
+func NewRecipient(broker Broker, val Validator, store TicketStore, secret [32]byte, faceValue *big.Int, winProb *big.Int) Recipient {
+	return &recipient{
+		broker:       broker,
+		val:          val,
+		store:        store,
+		secret:       secret,
```

Comment:
Perhaps we can generate `secret` inside of `Recipient`? Wondering why it's better to provide it externally?

Reply:
I though generating `secret` outside of `Recipient` would make testing easier especially in the future if we add a method for updating `secret`. Plus, the caller can choose how it wants to initialize `secret`.

Reply:
What concerns me is that in theory this design does not prevent the user of `Recipient` from reusing the same secret across multiple sessions. That is of course very dangerous because it exposes O to replays.

How do you suggest we think about this potential risk?

Reply:
So this is a question of whether the user should have to understand how to securely create a secret for this package. In the interest of making the package as mistake proof as possible I think creating the secret within `Recipient` makes sense. I guess I can always give `Recipient` a SecretGenerator interface that can facilitate stubbing in tests...

Reply:
Right. Then the risk moves to the `SecretGenerator`, where a pm package user might implement their own generator that also brings about potential security risks :)

I believe having at least one constructor function that does not take a `SecretGenerator` argument, but rather uses some default implementation is a good mitigation.

In addition, perhaps on the second constructor that does take in a generator, we should add very clear documentation that supplying your own generator adds risk.

wdyt?

Reply:
Updated with two constructors:

[NewRecipientWithSecret](https://github.com/livepeer/go-livepeer/pull/649/files#diff-3a62c7d0189200e6024095a62a267eaaR65) accepts a user provided secret.

[NewRecipient](https://github.com/livepeer/go-livepeer/pull/649/files#diff-3a62c7d0189200e6024095a62a267eaaR50) automatically generates a random secret and then uses NewRecipientWithSecret.